include/zephyr/bluetooth/gatt.h

```diff
@@ -165,21 +165,117 @@ typedef ssize_t (*bt_gatt_attr_write_func_t)(struct bt_conn *conn,
 					     const void *buf, uint16_t len,
 					     uint16_t offset, uint8_t flags);
 
-/** @brief GATT Attribute structure. */
+/** @brief GATT Attribute
+ *
+ *  This type primarily represents an ATT Attribute that may be
+ *  an entry in the local ATT database. The objects of this type
+ *  must be part of an array that forms a GATT service.
+ *
+ *  While the formed GATT service is registered with the local
+ *  GATT server, pointers to this type can typically be given to
+ *  GATT server APIs, like bt_gatt_notify().
+ *
+ *  @note This type is given as an argument to the
+ *  bt_gatt_discover() application callback, but it's not a
+ *  proper object of this type. The permissions, read() and
+ *  write() methods are not available, and it's unsound to pass
+ *  the pointer to GATT server APIs.
+ */
 struct bt_gatt_attr {
-	/** Attribute UUID */
+	/** @brief Attribute Type, aka. "UUID"
+	 *
+	 *  The Attribute Type determines the interface can be expected
+	 *  from the read() and write() methods and the possible
+	 *  permission configurations.
+	 *
+	 *  E.g. Attribute of type @ref BT_UUID_GATT_CPF will act as a
+	 *  GATT Characteristic Presentation Format descriptor as
+	 *  specified in Core Specification 3.G.3.3.3.5.
+	 *
+	 *  Anyone can define a new Attribute Type.
+	 */
 	const struct bt_uuid *uuid;
-	/** Attribute read callback */
+
+	/** @brief Attribute Value read method
+	 *
+	 *  This is the generic interface to read from the Attribute
+	 *  Value of this object in on-air format.
+	 *
+	 *  If this is non-NULL, you may call it to read the value.
+	 *  Value is the Attribute Value
+	 *
+	 *  This function may safely assume the Attribute Permissions
+	 *  are satisfied for this read.
+	 *
+	 *  The read() and write() implementations can and often do have
+	 *  side effects besides reading and writing the value.
+	 *
+	 *  This function is primarily invoked by the GATT server when a
+	 *  remote GATT client performs a read, but it can also be
+	 *  invoked for a local read.
+	 *
+	 *  Must be NULL if the attribute is not readable.
+	 */
 	bt_gatt_attr_read_func_t read;
```

Comment:
To some extend it's a bit weird that we document this callback both here and in the `typedef`. Users will need to read the documentation 2 places in order to understand how it's used and the arguments for it. 

This also applies for `write`

Reply:
Not sure what to do about this.

Reply:
Since it's technically an API change to remove the typedef and define the function directly here, then it may be better to add/move the newly added documentation to the typedef instead. 

We can also move the typedef documentation here, including the parameter and return value documentation

Reply:
I pushed a new version. I've moved text that I feel describes what the callback should do to the typedef, and I've kept text that talks about the pointer field itself here.